Solutions/Ais.Net.Receiver/NmeaReceiver.cs

```diff
@@ -2,99 +2,61 @@
 // Copyright (c) Endjin. All rights reserved.
 // </copyright>
 
-namespace Endjin.Ais.Receiver
+namespace Ais.Net.Receiver
 {
     using System;
     using System.Collections.Generic;
     using System.IO;
-    using System.Linq;
-    using System.Reactive.Subjects;
+    using System.Net.Sockets;
     using System.Threading.Tasks;
 
     public class NmeaReceiver
     {
-        private System.Net.Sockets.TcpClient tcpClient;
-        private Subject<string> items;
+        private readonly TcpClient tcpClient = new();
 
-        public NmeaReceiver(string host, int port)
+        public NmeaReceiver(string host, int port, TimeSpan? retryPeriodicity, int retryAttemptLimit = 100)
         {
-            Host = host;
-            Port = port;
-
-            this.tcpClient = new System.Net.Sockets.TcpClient();
+            this.Host = host;
+            this.Port = port;
+            this.RetryPeriodicity = (retryPeriodicity ??= TimeSpan.FromSeconds(1));
+            this.RetryAttemptLimit = retryAttemptLimit;
         }
 
-        public bool Connected
-        {
-            get { return this.tcpClient.Connected; }
-        }
+        public bool Connected => this.tcpClient.Connected;
 
         public string Host { get; }
 
         public int Port { get; }
 
-        public IObservable<string> Items
-        {
-            get { return this.items ?? (this.items = new Subject<string>()); }
-        }
+        public int RetryAttemptLimit { get; }
 
-        protected Subject<string> Subject
-        {
-            get { return this.items; }
-        }
+        public TimeSpan RetryPeriodicity { get;  }
 
-        public async Task InitaliseAsync()
+        public async IAsyncEnumerable<string> GetAsync()
         {
-            this.tcpClient = new System.Net.Sockets.TcpClient();
             await this.tcpClient.ConnectAsync(this.Host, this.Port);
-        }
+            await using NetworkStream stream = this.tcpClient.GetStream();
+            using StreamReader reader = new StreamReader(stream);
 
+            int retryAttempt = 0;
 
-        public async IAsyncEnumerable<string> GetAsync()
-        {
-            using (var reader = new StreamReader(this.tcpClient.GetStream()))
+            while (this.tcpClient.Connected)
             {
-                while (true)
+                while (stream.DataAvailable)
```

Comment:
I was staring baffled at this code for a while trying to understand how it handles retries. Initially I was expecting a `catch` block and retry logic in the event of failures. When I didn't find such a thing I wondered if maybe "retry" meant something else here—perhaps with this source we read all the data it has, and then it stops, and then to get the further data we have to wait a bit and then ask it if it's got anything new?

I don't think it's that.

My current provisional conclusion is that the failure mode we see in practice is that the connection gets closed, and so we fall out of this loop.

Anyway, in conclusion, I'd really like a comment here explaining what failure looks like.

Reply:
When I last tried to run the code in 2019, we would occasionally get into a state where the remote AIS Sender process would become unresponsive, but the host was still resolvable, and DataAvailable was the only mechanism for detecting that.